src/storage/src/upsert.rs

```diff
@@ -807,15 +749,36 @@ where
         }
 
         drop(events);
-
+        drop(permit);
         drop(previous_token);
+        drop(snapshot_cap);
         // Exchaust the previous input. It is expected to immediately reach the empty
         // antichain since we have dropped its token.
         //
-        // Note that we do not need to also process the `input` during this, as the dropped token
-        // will shutdown the `backpressure` operator
         while let Some(_event) = previous.next().await {}
 
+        // Drain the input to ensure we correctly process data from before the resume upper
+        'outer: while !PartialOrder::less_equal(&resume_upper, &input_upper) {
+            input.ready().await;
+            while let Some(input_event) = input.next_sync() {
```

Comment:
in this case you might as well call `input.next().await` directly and have a single loop